File: 05_action_server_client/src/action_server_client_python/action_server_client_python/action_client.py
import sys

import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node

from custom_action_interfaces.action import CustomAction

# 自定义动作客户端节点
class FirstActionClient(Node):
    def __init__(self):
        # 初始化节点，节点名为 custom_action_client_python
        super().__init__('custom_action_client_python')

        # 创建动作客户端：
        # 动作类型：CustomAction
        # 动作名称：custom_action_python_action
        self.action_client = ActionClient(
            self,
            CustomAction,
            'custom_action_python_action'
        )

    # 发送动作目标
    def send_goal(self, n):
        # 创建目标对象
        goal = CustomAction.Goal()

        # 设置目标参数
        goal.n = n

        # 等待动作服务端启动
        self.get_logger().info(
            '正在等待动作服务端启动...'
        )
        self.action_client.wait_for_server()

        # 异步发送目标：
        # goal               需要发送的目标对象
        # feedback_callback  接收到动作反馈时调用的回调函数
        self.send_goal_future = self.action_client.send_goal_async(
            goal,
            feedback_callback=self.feedback_callback
        )

        # 目标请求处理完成后，调用 goal_response_callback()
        self.send_goal_future.add_done_callback(
            self.goal_response_callback
        )

    # 目标响应回调函数
    def goal_response_callback(self, future):
        # 获取目标句柄
        goal_handle = future.result()

        # 判断动作服务端是否接受目标
        if not goal_handle.accepted:
            self.get_logger().info(
                '目标被拒绝'
            )
            rclpy.shutdown()
            return

        self.get_logger().info(
            '目标已接受，等待执行结果...'
        )

        # 异步请求最终执行结果
        self.get_result_future = goal_handle.get_result_async()

        # 获取结果后，调用 get_result_callback()
        self.get_result_future.add_done_callback(
            self.get_result_callback
        )

    # 动作反馈回调函数
    def feedback_callback(self, feedback_msg):
        # 获取反馈数据
        feedback = feedback_msg.feedback

        self.get_logger().info(
            f'收到反馈：{feedback.partial_sequence}'
        )

    # 动作结果回调函数
    def get_result_callback(self, future):
        # 获取动作结果
        result = future.result().result

        self.get_logger().info(
            f'收到结果：{result.sequence}'
        )

        # 等待键盘输入，暂缓执行后续代码，便于调试观察
        input()

        # 已经获得最终结果，关闭ROS 2上下文，从而终止spin()循环
        rclpy.shutdown()

def main(args=None):
    # 初始化 ROS 2 Python 客户端库
    rclpy.init(args=args)

    # 获取命令行参数
    n = int(sys.argv[1])

    # 创建动作客户端节点
    action_client_node = FirstActionClient()

    # 发送动作目标
    action_client_node.send_goal(n)

    # 保持节点运行，等待并处理目标响应、动作反馈和最终结果
    rclpy.spin(action_client_node)

    # 销毁节点
    action_client_node.destroy_node()

File: 05_action_server_client/src/action_server_client_cpp/src/action_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "custom_action_interfaces/action/custom_action.hpp"

#include <chrono>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <memory>
#include <sstream>

using namespace std::chrono_literals;

// 自定义动作客户端节点
class FirstActionClient : public rclcpp::Node
{
public:
    // 动作类型
    using CustomAction =
        custom_action_interfaces::action::CustomAction;

    // 动作客户端目标句柄类型
    using GoalHandleCustomAction =
        rclcpp_action::ClientGoalHandle<CustomAction>;

    FirstActionClient()
        // 初始化节点，节点名为 custom_action_client_cpp
        : Node("custom_action_client_cpp")
    {
        // 创建动作客户端：
        // 动作类型：CustomAction
        // 动作名称：custom_action_cpp_action
        action_client_ =
            rclcpp_action::create_client<CustomAction>(
                this,
                "custom_action_cpp_action"
            );
    }

    // 发送动作目标
    void send_goal(int32_t n)
    {
        // 创建目标对象
        CustomAction::Goal goal;

        // 设置目标参数
        goal.n = n;

        // 等待动作服务端启动
        RCLCPP_INFO(
            this->get_logger(),
            "正在等待动作服务端启动..."
        );

        while (!action_client_->wait_for_action_server(1s))
        {
            if (!rclcpp::ok())
            {
                RCLCPP_ERROR(
                    this->get_logger(),
                    "等待动作服务端时 ROS 2 已关闭。"
                );
                return;
            }
        }

        // 创建发送目标选项对象
        auto send_goal_options =
            rclcpp_action::Client<
                CustomAction>::SendGoalOptions();

        // 设置目标响应回调函数
        send_goal_options.goal_response_callback =
            std::bind(
                &FirstActionClient::goal_response_callback,
                this,
                std::placeholders::_1
            );

        // 设置动作反馈回调函数
        send_goal_options.feedback_callback =
            std::bind(
                &FirstActionClient::feedback_callback,
                this,
                std::placeholders::_1,
                std::placeholders::_2
            );

        // 设置动作结果回调函数
        send_goal_options.result_callback =
            std::bind(
                &FirstActionClient::get_result_callback,
                this,
                std::placeholders::_1
            );

        // 异步发送目标：
        // goal               需要发送的目标对象
        // send_goal_options  保存目标响应、反馈和结果回调函数
        action_client_->async_send_goal(
            goal,
            send_goal_options
        );
    }

private:
    // 目标响应回调函数
    void goal_response_callback(
        const GoalHandleCustomAction::SharedPtr &goal_handle)
    {
        // 空指针表示动作服务端拒绝了目标
        if (!goal_handle)
        {
            RCLCPP_INFO(
                this->get_logger(),
                "目标被拒绝"
            );

            rclcpp::shutdown();
            return;
        }

        RCLCPP_INFO(
            this->get_logger(),
            "目标已接受，等待执行结果..."
        );
    }

    // 动作反馈回调函数
    void feedback_callback(
        GoalHandleCustomAction::SharedPtr goal_handle,
        const std::shared_ptr<
            const CustomAction::Feedback> feedback)
    {
        // 本回调中不需要使用目标句柄
        (void)goal_handle;

        // 将反馈数据转换为字符串
        std::ostringstream feedback_stream;
        feedback_stream << "[";

        for (
            std::size_t i = 0;
            i < feedback->partial_sequence.size();
            ++i)
        {
            feedback_stream
                << feedback->partial_sequence[i];

            if (
                i
                < feedback->partial_sequence.size() - 1)
            {
                feedback_stream << ", ";
            }
        }

        feedback_stream << "]";

        // 输出反馈数据
        RCLCPP_INFO(
            this->get_logger(),
            "收到反馈：%s",
            feedback_stream.str().c_str()
        );
    }

    // 动作结果回调函数
    void get_result_callback(
        const GoalHandleCustomAction::WrappedResult &wrapped_result)
    {
        // 判断目标最终执行状态
        switch (wrapped_result.code)
        {
            case rclcpp_action::ResultCode::SUCCEEDED:
                break;

            case rclcpp_action::ResultCode::ABORTED:
                RCLCPP_ERROR(
                    this->get_logger(),
                    "目标执行被中止。"
                );

                rclcpp::shutdown();
                return;

            case rclcpp_action::ResultCode::CANCELED:
                RCLCPP_INFO(
                    this->get_logger(),
                    "目标已取消。"
                );

                rclcpp::shutdown();
                return;

            default:
                RCLCPP_ERROR(
                    this->get_logger(),
                    "收到未知的目标结果状态。"
                );

                rclcpp::shutdown();
                return;
        }

        // 获取动作结果
        const auto result = wrapped_result.result;

        // 将动作结果转换为字符串
        std::ostringstream result_stream;
        result_stream << "[";

        for (
            std::size_t i = 0;
            i < result->sequence.size();
            ++i)
        {
            result_stream << result->sequence[i];

            if (i < result->sequence.size() - 1)
            {
                result_stream << ", ";
            }
        }

        result_stream << "]";

        // 输出动作结果
        RCLCPP_INFO(
            this->get_logger(),
            "收到结果：%s",
            result_stream.str().c_str()
        );

        // 等待键盘输入，暂缓执行后续代码，便于调试观察
        std::cin.get();
        
        // 已经获得最终结果，关闭ROS 2上下文，终止spin()循环
        rclcpp::shutdown();
    }

    // 动作客户端对象
    rclcpp_action::Client<CustomAction>::SharedPtr
        action_client_;
};

int main(int argc, char **argv)
{
    // 初始化 ROS 2 C++ 客户端库
    rclcpp::init(argc, argv);

    // 检查命令行参数
    if (argc != 2)
    {
        RCLCPP_ERROR(
            rclcpp::get_logger("custom_action_client_cpp"),
            "用法：custom_action_client_cpp N"
        );

        rclcpp::shutdown();
        return 1;
    }

    // 获取命令行参数
    int32_t n = std::atoi(argv[1]);

    // 创建动作客户端节点
    auto action_client_node =
        std::make_shared<FirstActionClient>();

    // 发送动作目标
    action_client_node->send_goal(n);

    // 保持节点运行，
    // 等待并处理目标响应、动作反馈和最终结果
    rclcpp::spin(action_client_node);

    // 销毁节点并关闭 ROS 2
    action_client_node.reset();

    return 0;
}
